src/main/java/week6/ParkingLot.java

```diff
@@ -0,0 +1,153 @@
+  private static final Logger logger = Logger.getLogger(ParkingLot.class.getName());
+
+  /*
+   * First, we create a map of wrongly positioned cars, car no. -> position
+   * Iteratively, replace 0 with the correct car. We remove cars that are in correct pos from map
+   *
+   * If 0 becomes in the correct position but map is not empty, move 0 to first incorrect car pos
+   * and start iterating again until map is empty.
+   *
+   * Amortized Time Complexity: O(n)
+   * Space Complexity: O(n)
+   *
+   * where n is the size of the array source
+   *
+   * Assumptions:
+   *    1) len(source) = len(target)
+   *    2) arrays contain O...n, this means no duplications
+   *
+   * @param source : the initial configuration of the cars
+   * @param target : the target configuration of the cars
+   * @return : List of Moves to be performed to transform source -> target
+   */
+  public static List<Move> rearrangeCars(final int[] source, final int[] target) {
+
+    List<Move> moves = new ArrayList<Move>() {
+      @Override
+      public String toString() {
+        return super.toString().replaceAll("\\[|]|(, )", "");
+      }
+    };
+
+    if (source == null || target == null || source.length != target.length) {
+      return moves;
+    }
+
+    //creates map (wrongly positioned) car id -> position and get index of the empty Box
+    HashMap<Integer, Integer> misplacedCars = new HashMap<>();
+    int emptyBox = populateMap(source, target, misplacedCars);
+
+    while (!misplacedCars.isEmpty()) {
+      //if emptyBox is in right pos, we swap the emptyBox with one of the wrong cars
+      emptyBox = correctEmpty(misplacedCars, target, emptyBox, moves);
+      emptyBox = moveEmpty(misplacedCars, target, emptyBox, moves);
+    }
+
+    logger.info(moves.toString());
+    return moves;
+  }
+
+
+  /*
+   * Updates the map misPlacedCars : cars that needs to be moved -> current position.   *
+   * Returns the index of the empty box
+   *
+   * Time Complexity: O(n) where n is the size of array source
+   * Space Complexity: O(1)
+   *
+   * @param source : the initial configuration of the cars
+   * @param target : the target configuration of the cars
+   * @param misplacedCars : the map to be updated cars to be moved -> their current position
+   * @return : the index of the empty box
+   */
+  private static int populateMap(int[] source, int[] target,
+      HashMap<Integer, Integer> misplacedCars) {
+    int emptyBox = -1;
+    for (int i = 0; i < source.length; i++) {
+      if (source[i] == EMPTY) {
+        emptyBox = i;
+      } else if (source[i] != target[i]) {
+        misplacedCars.put(source[i], i);
+      }
+    }
+    return emptyBox;
+  }
+
+  /*
+  * This steps is required if emptyBox is in the right position but the list is not sorted.
+  *     e.g. cars = [1, 0, 2, 3], target = [3, 0, 2, 1] and emptyBox = 1
+  * We swap the emptyBox with one of the cars in the map (wrongly positioned), here we perform
+  *     "move from 1 to 0"
```

Comment:
Shouldn't this be "move from 0 to 1"? Because according to the given example, if we're moving the car with the number 1 when the arrangement is [1, 0, 2, 3], then the move would be described as "move from 0 to 1"